src/downloader.rs

```diff
@@ -91,20 +92,58 @@ impl Downloader {
         main: Arc<ProgressBar>,
     ) -> Summary {
         // Create a download summary.
-        let mut summary = Summary::new(download.clone(), StatusCode::BAD_REQUEST, 0);
-
-        // Check if we should skip the file or continue.
+        let mut size_on_disk: u64 = 0;
+        let mut resumable = false;
+        let mut summary = Summary::new(
+            download.clone(),
+            StatusCode::BAD_REQUEST,
+            size_on_disk,
+            resumable,
+        );
         let output = self.directory.join(&download.filename);
-        if self.skip_existing && output.exists() {
-            return summary.with_status(Status::Skipped(format!(
-                "a file with the same name already exists at the destination: {:?}",
-                &output
-            )));
+
+        // If resumable is turned on...
+        if self.resumable {
+            // Check whether the download is resumable.
+            let res = match client.head(download.url.clone()).send().await {
+                Ok(res) => res,
+                Err(e) => {
+                    return summary.fail(e);
+                }
+            };
+            let headers = res.headers();
+            resumable = match headers.get(ACCEPT_RANGES) {
+                None => false,
+                Some(x) if x == "none" => false,
+                Some(_) => true,
+            };
+
+            // If we can resume...
+            if resumable {
+                // Check if there is a file on disk already.
+                if output.exists() {
+                    debug!("A file with the same name already exists at the destination.");
+                    // If so, check file length to know where to restart the download from.
+                    size_on_disk = match output.metadata() {
+                        Ok(m) => m.len(),
+                        Err(e) => {
+                            return summary.fail(e);
+                        }
+                    }
+                }
+            }
+
+            // Update the summary accordingly.
+            summary.set_resumable(resumable);
         }
 
         // Request the file.
         debug!("Fetching {}", &download.url);
-        let res = match client.get(download.url.clone()).send().await {
+        let mut req = client.get(download.url.clone());
+        if self.resumable && resumable {
```

Comment:
should one of these variables be renamed? it seems tricky to reason `if self.think && thing`

Reply:
Yeah, same here, I can try to make this part more readable.